File: aws-sdk-java/LeoFSTest.java
// Java Built-in
import java.io.IOException;
import java.io.File;
import java.io.FileInputStream;
import java.io.BufferedInputStream;
import java.io.RandomAccessFile;
import java.security.MessageDigest;
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

// AWS-SDK-Java
import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.auth.AWSCredentials;
import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.ClientConfiguration;
import com.amazonaws.Protocol;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3Client;
import com.amazonaws.services.s3.S3ClientOptions;
import com.amazonaws.services.s3.model.AccessControlList;
import com.amazonaws.services.s3.model.AmazonS3Exception;
import com.amazonaws.services.s3.model.Bucket;
import com.amazonaws.services.s3.model.CannedAccessControlList;
import com.amazonaws.services.s3.model.DeleteObjectsRequest;
import com.amazonaws.services.s3.model.DeleteObjectsResult;
import com.amazonaws.services.s3.model.DeleteObjectsResult.DeletedObject;
import com.amazonaws.services.s3.model.GetObjectRequest;
import com.amazonaws.services.s3.model.Grant;
import com.amazonaws.services.s3.model.ListObjectsRequest;
import com.amazonaws.services.s3.model.ObjectListing;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.PutObjectRequest;
import com.amazonaws.services.s3.model.S3Object;
import com.amazonaws.services.s3.model.S3ObjectSummary;
import com.amazonaws.services.s3.transfer.TransferManager;
import com.amazonaws.services.s3.transfer.Upload;

// Extra
import org.apache.commons.io.IOUtils;

public class LeoFSTest {
    private static String host = "localhost";   // leo_gateway host
    private static Integer port = 8080;         // leo_gateway port

    private static String accessKeyId       = "05236";
    private static String secretAccessKey   = "802562235";

    private static String signVer   = "v4";

    private static String bucket    = "testj";
    private static String tempData  = "../temp_data/";

    private static final String smallTestF  = tempData + "testFile";
    private static final String mediumTestF  = tempData + "testFile.medium";
    private static final String largeTestF  = tempData + "testFile.large"; 

    private static String metadataKey = "cmeta_key";
    private static String metadataVal = "cmeta_val";
    private static Map<String, String> metadataMap;

    private static ClientConfiguration config;
    private static AWSCredentials credentials;
    private static AmazonS3 s3;


    public static void main(String[] args) throws IOException, InterruptedException {
        if (args.length > 0)
            signVer = args[0];
        if (args.length > 1) {
			host = args[1];
			port = Integer.parseInt(args[2]);
            bucket = args[3];
		}

        System.out.println(signVer + ", " + bucket);
        // Init
        init(signVer);
        createBucket(bucket);

        // Put Object Test
        putObject(bucket, "test.simple",    smallTestF);
        putObject(bucket, "test.medium",    mediumTestF);
        putObject(bucket, "test.large",     largeTestF);

        // Put Object with Metadata Test
        putObjectWithMetadata(bucket, "test.simple.meta", smallTestF, metadataMap);
        putObjectWithMetadata(bucket, "test.large.meta", largeTestF, metadataMap);

        // Multipart Upload Test
        mpObject(bucket, "test.simple.mp",  smallTestF);
        mpObject(bucket, "test.large.mp",   largeTestF);

        // Object Metadata Test
        headObject(bucket, "test.simple",   smallTestF);
        headObject(bucket, "test.simple.mp", smallTestF);
        headObject(bucket, "test.large",    largeTestF);
// MP File ETag != MD5
//        headObject(bucket, "test.large.mp", largeTestF);

        // Get Object Test
        getObject(bucket, "test.simple",    smallTestF);
        getObject(bucket, "test.simple.mp", smallTestF);
        getObject(bucket, "test.medium",    mediumTestF);
        getObject(bucket, "test.large",     largeTestF);
// MP File ETag != MD5
//        getObject(bucket, "test.large.mp",  largeTestF);

        // Get Objetc Again (Cache) Test
        getObject(bucket, "test.simple",    smallTestF);
        getObject(bucket, "test.simple.mp", smallTestF);
        getObject(bucket, "test.medium",    mediumTestF);
        getObject(bucket, "test.large",     largeTestF);

        // Get Object with Metadata Test
        getObjectWithMetadata(bucket, "test.simple.meta", smallTestF, metadataMap);
        getObjectWithMetadata(bucket, "test.large.meta", largeTestF, metadataMap);

        // Get Not Exist Object Test
        getNotExist(bucket, "test.noexist");

        // Range Get Object Test
        rangeObject(bucket, "test.simple",      smallTestF, 1, 4); 
        rangeObject(bucket, "test.simple.mp",   smallTestF, 1, 4); 
        rangeObject(bucket, "test.large",       largeTestF, 1048576, 10485760); 
        rangeObject(bucket, "test.large.mp",    largeTestF, 1048576, 10485760); 

        // Copy Object Test
        copyObject(bucket, "test.simple", "test.simple.copy");
        getObject(bucket, "test.simple.copy", smallTestF);

        // List Object Test
        listObject(bucket, "", -1);

        // Delete All Object Test
        deleteAllObjects(bucket);
        Thread.sleep(3000); // Need to wait for syncing directory metadata
        listObject(bucket, "", 0);

        // Multiple Page List Object Test
        putDummyObjects(bucket, "list/", 35, smallTestF);
        Thread.sleep(3000); // Need to wait for syncing directory metadata
        pageListBucket(bucket, "list/", 35, 10);

        // Multiple Delete
        multiDelete(bucket, "list/", 10);

        // GET-PUT ACL
        setBucketAcl(bucket, "private");
        setBucketAcl(bucket, "public-read");
        setBucketAcl(bucket, "public-read-write");
    }

    public static void init(String SignVer) {
        System.out.println("----- Init Start -----");
        config = new ClientConfiguration();
        config.setProxyHost(host);
        config.setProxyPort(port);
        config.withProtocol(Protocol.HTTP);
        if (SignVer.equals("v4")) {
            config.setSignerOverride("AWSS3V4SignerType");
        }
        credentials = new BasicAWSCredentials(accessKeyId, secretAccessKey);
        s3 = new AmazonS3Client(credentials, config);
        S3ClientOptions opts = new S3ClientOptions();
        metadataMap = new HashMap();
        metadataMap.put(metadataKey, metadataVal);
        opts.setPathStyleAccess(true);
        s3.setS3ClientOptions(opts);
        System.out.println("----- Init End -----");
        System.out.println();
    }

    public static void createBucket(String bucketName) throws IOException {
        System.out.println("===== Create Bucket [" + bucketName + "] Start =====");
        s3.createBucket(bucketName);
        if (!s3.doesBucketExist(bucketName)) {
            throw new IOException("Create Bucket [" + bucketName + "] Failed!");
        }
        System.out.println("===== Create Bucket End =====");
        System.out.println();
    }

    private static void doPutObject(String bucketName, String key, String path) {
        File file = new File(path);
        s3.putObject(new PutObjectRequest(bucketName, key, file));
    }

    private static void doPutObject(String bucketName, String key, String path, Map<String, String> meta_map) throws IOException {
        BufferedInputStream bufferedStream = new BufferedInputStream(new FileInputStream(path));
        File file = new File(path);
        PutObjectRequest req = new PutObjectRequest(bucketName, key, file);
        ObjectMetadata meta = new ObjectMetadata();
        meta.setUserMetadata(meta_map);
        req.withMetadata(meta);
        s3.putObject(req);
    }

    public static void putObject(String bucketName, String key, String path) throws IOException {
        System.out.println("===== Put Object [" + bucketName + "/" + key + "] Start =====");
        doPutObject(bucketName, key, path);
        if (!doesFileExist(bucketName, key)) {
            throw new IOException("Put Object [" + bucketName + "/" + key + "] Failed!");
        }
        System.out.println("===== Put Object End =====");
        System.out.println();
    }

    public static void putObjectWithMetadata(String bucketName, String key, String path, Map<String, String> meta_map) throws IOException {
        System.out.println("===== Put Object [" + bucketName + "/" + key + "] with Metadata Start =====");
        doPutObject(bucketName, key, path, meta_map);
        if (!doesFileExist(bucketName, key)) {
            throw new IOException("Put Object [" + bucketName + "/" + key + "] with Metadata Failed!");
        }
        System.out.println("===== Put Object with Metadata End =====");
        System.out.println();
    }

    public static void mpObject(String bucketName, String key, String path) throws IOException, InterruptedException {
        System.out.println("===== Multipart Upload Object [" + bucketName + "/" + key + "] Start =====");
        TransferManager tx = new TransferManager(s3);
        BufferedInputStream bufferedStream = new BufferedInputStream(new FileInputStream(path));

        File file = new File(path);
        long fileSize = file.length();
        ObjectMetadata meta = new ObjectMetadata();
        meta.setContentLength(fileSize);
        Upload upload = tx.upload(bucketName, key, bufferedStream, meta); 
        while (upload.isDone() == false) {
            System.out.println(" - Progress: " + upload.getProgress().getBytesTransferred() 
                                  + "Byte \t" + upload.getProgress().getPercentTransferred() + "%" );
            Thread.sleep(100);
        } 
        upload.waitForCompletion();
        tx.shutdownNow(Boolean.FALSE.booleanValue());
        if (!doesFileExist(bucketName, key)) {
            throw new IOException("Multipart Upload Object [" + bucketName + "/" + key + "] Failed!");
        }
        System.out.println("===== Multipart Upload Object End =====");
        System.out.println();
    }

    public static void headObject(String bucketName, String key, String path) throws IOException {
        System.out.println("===== Head Object [" + bucketName + "/" + key + "] Start =====");
        ObjectMetadata meta = s3.getObjectMetadata(bucketName, key);
        String etag = meta.getETag();
        long length = meta.getContentLength();
        File file = new File(path);
        String md5 = MD5(path);
        System.out.println("ETag: " + etag + ", Size: " + length);
        if (file.length() != length || !md5.equals(etag)) {
            throw new IOException("Metadata [" + bucketName + "/" + key + "] NOT Match, Size: " + file.length() + ", MD5: " + md5);
        }
        System.out.println("===== Head Object End =====");
        System.out.println();
    }

    public static void getObject(String bucketName, String key, String path) throws IOException {
        System.out.println("===== Get Object [" + bucketName + "/" + key + "] Start =====");
        S3Object object = s3.getObject(bucketName, key);
        if (!IOUtils.contentEquals(object.getObjectContent(), new FileInputStream(path))) {
            throw new IOException("Content NOT Match!");
        }
        System.out.println("===== Get Object End =====");
        System.out.println();
    }

    public static void getObjectWithMetadata(String bucketName, String key, String path, Map<String, String> meta_map) throws IOException {
        System.out.println("===== Get Object [" + bucketName + "/" + key + "] with Metadata Start =====");
        S3Object object = s3.getObject(bucketName, key);
        Map<String, String> meta = object.getObjectMetadata().getUserMetadata();
        if (!meta.equals(meta_map)) {
            throw new IOException("Metadata NOT Match!");
        }
        if (!IOUtils.contentEquals(object.getObjectContent(), new FileInputStream(path))) {
            throw new IOException("Content NOT Match!");
        }
        System.out.println("===== Get Object with Metadata End =====");
        System.out.println();
    }

    public static void rangeObject(String bucketName, String key, String path, int start, int end) throws IOException {
        System.out.println("===== Range Get Object [" + bucketName + "/" + key + "] (" + start + "-" + end + ") Start =====");
        S3Object object = s3.getObject(new GetObjectRequest(bucketName, key).withRange(start, end));
        byte [] res = IOUtils.toByteArray(object.getObjectContent());

        RandomAccessFile raf = new RandomAccessFile(path, "r");
        raf.seek(start);
        byte [] tmp = new byte[end - start + 1];
        raf.readFully(tmp);

        if (!Arrays.equals(res, tmp)) {
            throw new IOException("Content NOT Match!");
        }
        System.out.println("===== Range Get Object End =====");
        System.out.println();
    }

    public static void getNotExist(String bucketName, String key) throws IOException {
        System.out.println("===== Get Not Exist Object [" + bucketName + "/" + key + "] Start =====");
        try {
            S3Object object = s3.getObject(bucketName, key);
            throw new IOException("Should NOT Exist!");
        } catch (AmazonS3Exception s3e) {
            int code = s3e.getStatusCode();
            if (code != 403 && code != 404) {
                throw new IOException("Incorrect Status Code [" + code + "]!");
            }
        }
        System.out.println("===== Get Not Exist Object End =====");
        System.out.println();
    }

    public static void copyObject(String bucketName, String src, String dst) throws IOException {
        System.out.println("===== Copy Object [" + 
                bucketName + "/" + src + "] -> [" + 
                bucketName + "/" + dst + "] Start =====");
        s3.copyObject(bucketName, src, bucketName, dst);
        if (!doesFileExist(bucketName, dst)) {
            throw new IOException("Copy Object Failed!");
        }
        System.out.println("===== Copy Object End =====");
        System.out.println();
    }

    public static void listObject(String bucketName, String prefix, int expected) throws IOException {
        System.out.println("===== List Objects [" + bucketName + "/" + prefix + "*] Start =====");
        ObjectListing objList = s3.listObjects(bucketName, prefix);
        int count = 0;
        for (S3ObjectSummary objectSummary : objList.getObjectSummaries()) {
            if (doesFileExist(bucketName, objectSummary.getKey())) {
                System.out.println(objectSummary.getKey() + " \t Size: " + objectSummary.getSize());
                count ++;
            }
        }
        if (expected >= 0) {
            if (count != expected) {
                throw new IOException("Number of Objects NOT Match!");
            }
        }

        System.out.println("===== List Objects End =====");
        System.out.println();
    }

    public static void deleteAllObjects(String bucketName) throws IOException {
        System.out.println("===== Delete All Objects [" + bucketName + "] Start =====");
        ObjectListing objList = s3.listObjects(bucketName, "");
        for (S3ObjectSummary objectSummary : objList.getObjectSummaries()) {
            s3.deleteObject(bucketName, objectSummary.getKey());
        }

        System.out.println("===== Delete All Objects End =====");
        System.out.println();
    }

    public static void putDummyObjects(String bucketName, String prefix, int total, String holder) throws IOException {
        for (int i = 0; i < total; i++) {
            doPutObject(bucketName, prefix + i, holder);
        }
    }

    public static void pageListBucket(String bucketName, String prefix, int total, int pageSize) throws IOException {
        System.out.println("===== Multiple Page List Objects [" + bucketName + "/" + prefix + "*] " + total + " Objs @" + pageSize + " Start =====");
        ObjectListing objList = s3.listObjects(new ListObjectsRequest(bucketName, prefix, null, null, pageSize));
        int actualCount = 0;
        while(true){
            System.out.println("===== Page =====");
            for (S3ObjectSummary objectSummary : objList.getObjectSummaries()) {
                actualCount++;
                System.out.println(objectSummary.getKey() 
                        + " \t Size: " + objectSummary.getSize()
                        + " \t Count: " + actualCount);
            }
            if (!objList.isTruncated()) {
                break;
            } else {
                objList.setBucketName(bucketName);
                objList = s3.listNextBatchOfObjects(objList);
            }
        }
        System.out.println("===== End =====");
        if (total != actualCount) {
            throw new IOException("Number of Objects NOT Match!");
        }
        System.out.println("===== Multiple Page List Objects End =====");
        System.out.println();
    }

    public static void multiDelete(String bucketName, String prefix, int total) throws IOException{
        System.out.println("===== Multiple Delete Objects [" + bucketName + "/" + prefix + "] Start =====");
        List<String> delList = new ArrayList<String>();
        for (int i = 0; i < total; i++) { 
            delList.add(prefix + i);
        }
        String [] delKeys = delList.toArray(new String[total]);
        DeleteObjectsResult delRes = s3.deleteObjects(new DeleteObjectsRequest(bucketName).withKeys(delKeys));
        List<DeleteObjectsResult.DeletedObject> delObjList = delRes.getDeletedObjects();
        for (DeleteObjectsResult.DeletedObject delObj : delObjList) {
            System.out.println("Deleted " + bucketName + "/" + delObj.getKey());
        }
        if (delObjList.size() != total) {
            throw new IOException("Number of Objects NOT Match!");
        }

        System.out.println("===== Multiple Delete Objects End =====");
        System.out.println();
    }

    public static void setBucketAcl(String bucketName, String permission) throws IOException {
        System.out.println("===== Set Bucket ACL [" + bucketName + "] (" + permission + ") Start =====");
        CannedAccessControlList targetAcl;
        List<String> checkList = new ArrayList<String>();
        if (permission.equals("private")) {
            targetAcl = CannedAccessControlList.Private;
            checkList.add("FULL_CONTROL");
        } else if (permission.equals("public-read")) {
            targetAcl = CannedAccessControlList.PublicRead;
            checkList.add("READ");
            checkList.add("READ_ACP");
        } else if (permission.equals("public-read-write")) {
            targetAcl = CannedAccessControlList.PublicReadWrite;
            checkList.add("READ");
            checkList.add("READ_ACP");
            checkList.add("WRITE");
            checkList.add("WRITE_ACP");
        } else {
            throw new IOException("Invalid Permission!");
        }
        s3.setBucketAcl(bucketName, targetAcl);
        AccessControlList acl = s3.getBucketAcl(bucketName);
        System.out.println("Owner ID: " + acl.getOwner());
        List<String> list = new ArrayList<String>();
        for (Grant grant : acl.getGrants()) {
            System.out.println("Grantee : " + grant.getGrantee() + " \t Permissions: " + grant.getPermission()); 
            list.add(grant.getPermission().toString());
        }
        for (String checkItem : checkList) {
            if (!list.contains(checkItem)) {
                throw new IOException("ACL NOT Match!");
            }
        }

        System.out.println("===== Set Bucket ACL End =====");
        System.out.println();
    }

    public static boolean doesFileExist(String bucketName, String key ) throws AmazonClientException{
        boolean isValidFile = true;
        try {
            ObjectMetadata objectMetadata = s3.getObjectMetadata(bucketName, key);
        } catch ( AmazonS3Exception s3e ) {
            if ( s3e.getStatusCode() == 404 ) {
                // i.e. 404: NoSuchKey - The specified key does not exist
                isValidFile = false;
            }
            else {
                throw s3e; // rethrow all S3 exceptions other than 404 
            }
        }
        catch ( Exception exception ) {
            exception.printStackTrace();
            isValidFile = false;
        }
        return isValidFile;
    }

    public static String MD5( String filePath )
    {
        StringBuffer sb = new StringBuffer();
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            FileInputStream fis = new FileInputStream(filePath); 
            byte[] dataBytes = new byte[1024];

            int nread = 0; 
            while ( (nread = fis.read(dataBytes)) != -1 ) {
                md.update(dataBytes, 0, nread);
            };
            byte[] mdbytes = md.digest();

            //convert the byte to hex format method 1
            for (int i = 0; i < mdbytes.length; i++) {
                sb.append(Integer.toString((mdbytes[i] & 0xff) + 0x100, 16).substring(1));
            }
 
        } catch ( java.security.NoSuchAlgorithmException e ) {
            System.out.println(e.getMessage());
        } catch ( Exception e ) {
            System.out.println(e.getMessage());
        }
        return sb.toString();
    }
}

File: boto/LeoFSTest.py
#!/usr/bin/python
# coding: utf8

from boto.s3.connection import S3Connection, OrdinaryCallingFormat
from boto.s3.bucket import Bucket
from boto.s3.key import Key
from filechunkio import FileChunkIO
import boto
import traceback
import os
import hashlib
import sys
from functools import partial

Host    = "localhost"
Port    = 8080

AccessKeyId     = "05236"
SecretAccessKey = "802562235"
SignVer         = "v4"

Bucket      = "testb"
TempData    = "../temp_data/"

SmallTestF  = TempData + "testFile"
LargeTestF  = TempData + "testFile.large"

s3 = None

def main():
    global SignVer
    if len(sys.argv) > 1:
        SignVer = sys.argv[1]
    try:
        init(SignVer)
        createBucket(Bucket)

        # Put Object Test
        putObject(Bucket, "test.simple",    SmallTestF)
        putObject(Bucket, "test.large",     LargeTestF)
    
#        # Multipart Upload Object Test
#        mpObject(Bucket, "test.simple.mp",  SmallTestF)
#        mpObject(Bucket, "test.large.mp",   LargeTestF)
    
        # Head Object Test
        headObject(Bucket, "test.simple",   SmallTestF)
        headObject(Bucket, "test.large",    LargeTestF)
#        headObject(Bucket, "test.simple.mp",SmallTestF)
    
        # Get Object Test
        getObject(Bucket, "test.simple",    SmallTestF)
        getObject(Bucket, "test.large",     LargeTestF)
#        getObject(Bucket, "test.simple.mp", SmallTestF)
#        getObject(Bucket, "test.large.mp",  LargeTestF)
    
        # Get Not Exist Object Test
        getNotExist(Bucket, "test.noexist")
    
        # Range Get Object Test
        rangeObject(Bucket, "test.simple",      SmallTestF, 1, 4)
        rangeObject(Bucket, "test.large",       LargeTestF, 1048576, 10485760)
#        rangeObject(Bucket, "test.simple.mp",   SmallTestF, 1, 4)
#        rangeObject(Bucket, "test.large.mp",    LargeTestF, 1048576, 10485760)
    
        # Copy Object Test
        copyObject(Bucket, "test.simple", "test.simple.copy")
        getObject(Bucket, "test.simple.copy", SmallTestF)
    
        # List Object Test
        listObject(Bucket, "", -1)
    
        # Delete All Object Test
        deleteAllObjects(Bucket)
        listObject(Bucket, "", 0)
    
        # Multiple Page List Object Test
        putDummyObjects(Bucket, "list/", 35, SmallTestF)
        pageListBucket(Bucket, "list/", 35, 10)
    
        # Multiple Delete
        multiDelete(Bucket, "list/", 10)
    
        # GET-PUT ACL
        setBucketAcl(Bucket, "private")
        setBucketAcl(Bucket, "public-read")
        setBucketAcl(Bucket, "public-read-write")

    except Exception, e:
        print traceback.format_exc()

def init(signVer):
    global s3
    if signVer == "v4":
        boto.config.add_section('s3')
        boto.config.set('s3', 'use-sigv4', 'True')
    else:
        boto.config.add_section('s3')
        boto.config.set('s3', 'use-sigv4', 'False')
    s3 = S3Connection(
            AccessKeyId,
            SecretAccessKey,
            host = 's3.amazonaws.com',
            proxy = Host,
            proxy_port = Port,
            calling_format = OrdinaryCallingFormat(),
            is_secure = False
            )

def createBucket(bucketName):
    print "===== Create Bucket [%s] Start =====" % bucketName
    s3.create_bucket(bucketName)
    print "===== Create Bucket End ====="
    print 

def putObject(bucketName, key, path):
    print "===== Put Object [%s/%s] Start =====" % (bucketName, key)
    bucket = s3.get_bucket(bucketName, validate=False)
    bucket.new_key(key).set_contents_from_filename(path)
    if not doesFileExist(bucketName, key):
        raise ValueError("Put Object [%s/%s] Failed!" % (bucketName, key))
    print "===== Put Object End ====="
    print

def mpObject(bucketName, key, path):
    print "===== Multipart Upload Object [%s/%s] Start =====" % (bucketName, key)
    bucket = s3.get_bucket(bucketName, validate=False)
    mp = bucket.initiate_multipart_upload(key)
    fileSize = os.path.getsize(path)
    offset = 0
    count = 0

    while (fileSize > offset):
        partSize = min(5 * 1024 * 1024, fileSize - offset)
        count = count + 1
        with FileChunkIO(path, 'rb', offset = offset, bytes=partSize) as fp:
            mp.upload_part_from_file(fp, part_num = count)
        offset = offset + partSize
    mp.complete_upload()

    if not doesFileExist(bucketName, key):
        raise ValueError("Multipary Upload Object [%s/%s] Failed!" % (bucketName, key))
    print "===== Multipart Upload Object End ====="
    print

def headObject(bucketName, key, path):
    print "===== Head Object [%s/%s] Start =====" % (bucketName, key)
    bucket = s3.get_bucket(bucketName, validate=False)
    obj = bucket.get_key(key)
    size = os.path.getsize(path)
    context = hashlib.md5()
    with open(path, 'rb') as f:
        for chunk in iter(partial(f.read, 4096), ''):
            context.update(chunk)
    md5sum = context.hexdigest()
    etag = obj.etag[1:-1]

    print "ETag: %s, Size: %d" % (etag, obj.size)
    if etag != md5sum or obj.size != size:
        raise ValueError("Metadata [%s/%s] NOT Match, Size: %d, MD5: %s" % (bucketName, key, size, md5sum))
    print "===== Head Object End ====="
    print

def getObject(bucketName, key, path):
    print "===== Get Object [%s/%s] Start =====" % (bucketName, key)
    bucket = s3.get_bucket(bucketName, validate=False)
    obj = bucket.get_key(key)
    file = open(path)
    if not doesFileMatch(file, obj):
        raise ValueError("Content NOT Match!")
    print "===== Get Object End ====="
    print

def getNotExist(bucketName, key):
    print "===== Get Not Exist Object [%s/%s] Start =====" % (bucketName, key)
    try:
        obj.read(1)
        raise ValueError("Should NOT Exist!")
    except Exception, not_found:
        pass
    print "===== Get Not Exist Object End ====="
    print

def rangeObject(bucketName, key, path, start, end):
    print "===== Range Get Object [%s/%s] (%d-%d) Start =====" % (bucketName, key, start, end)
    bucket = s3.get_bucket(bucketName, validate=False)
    obj = bucket.get_key(key)
    obj.open_read(headers = {'Range': 'bytes=%d-%d' % (start, end)})
    file = FileChunkIO(path, 'rb', offset=start, bytes=end - start + 1)
    if not doesFileMatch(file, obj):
        raise ValueError("Content NOT Match!")
    print "===== Range Get Object End ====="
    print

def copyObject(bucketName, src, dst):
    print "===== Copy Object [%s/%s] -> [%s/%s] Start =====" % (bucketName, src, bucketName, dst)
    bucket = s3.get_bucket(bucketName, validate=False)
    bucket.copy_key(dst, bucketName, src)
    print "===== Copy Object End ====="
    print

def listObject(bucketName, prefix, expected):
    print "===== List Objects [%s/%s*] Start =====" % (bucketName, prefix)
    bucket = s3.get_bucket(bucketName, validate=False)
    count = 0
    for obj in bucket.list():
        if doesFileExist(bucketName, obj.key):
            print "%s \t Size: %d\n" % (obj.key, obj.size)
            count = count + 1
    if expected >= 0 and count != expected:
        raise ValueError("Number of Objects NOT Match!")
    print "===== List Objects End ====="
    print

def deleteAllObjects(bucketName):
    print "===== Delete All Objects [%s] Start =====" % bucketName
    bucket = s3.get_bucket(bucketName, validate=False)
    for obj in bucket.list():
        obj.delete()
    print "===== Delete All Objects End ====="
    print

def putDummyObjects(bucketName, prefix, total, holder):
    bucket = s3.get_bucket(bucketName, validate=False)
    for i in range(0, total):
        bucket.new_key(prefix+str(i)).set_contents_from_filename(holder)

def pageListBucket(bucketName, prefix, total, pageSize):
    print "===== Multiple Page List Objects [%s/%s*] %d Objs @%d Start =====" % (bucketName, prefix, total, pageSize)
    bucket = s3.get_bucket(bucketName, validate=False)
    marker = ""
    count = 0
    while True:
        print "===== Page ====="
        res = bucket.get_all_keys(
                prefix = prefix,
                max_keys = pageSize,
                marker = marker
                )
        for obj in res:
            count = count + 1
            print "%s \t Size: %d \t Count: %d" % (obj.key, obj.size, count)
        if not res.is_truncated:
            break
        else:
            marker = res.next_marker
    print "===== End ====="
    if count != total:
        raise ValueError("Number of Objects NOT Match!")
    print "===== Multiple Page List Objects End ====="
    print

def multiDelete(bucketName, prefix, total):
    print "===== Multiple Delete Objects [%s/%s] Start =====" % (bucketName, prefix)
    bucket = s3.get_bucket(bucketName, validate=False)
    delKeyList = []
    for i in range(0, total):
        delKeyList.append(prefix+str(i))
    res = bucket.delete_keys(delKeyList)
    for obj in res.deleted:
        print "Deleted %s/%s" % (bucketName, obj.key)
    if len(res.deleted) != total:
        raise ValueError("Number of Objects NOT Match!")
    print "===== Multiple Delete Objects End ====="
    print

def setBucketAcl(bucketName, permission):
    print "===== Set Bucket ACL [%s] (%s) Start =====" % (bucketName , permission)
    if permission == "private":
        checkList = ["FULL_CONTROL"]
    elif permission == "public-read":
        checkList = ["READ", "READ_ACP"]
    elif permission == "public-read-write":
        checkList = ["READ", "READ_ACP", "WRITE", "WRITE_ACP"]
    else:
        raise ValueError("Invalid Permission!")
    bucket = s3.get_bucket(bucketName, validate=False)
    bucket.set_acl(permission)
    res = bucket.get_acl()
    print "Owner ID: S3Owner [name=%s,id=%s]" % (res.owner.display_name, res.owner.id)
    list = []
    for grant in res.acl.grants:
        print "Grantee: %s \t Permissions: %s" % (grant.uri, grant.permission)
        list.append(grant.permission)
    if not all(x in list for x in checkList):
        raise ValueError("ACL NOT Match!")
    print "===== Set Bucket ACL End ====="
    print

def doesFileExist(bucketName, key):
    bucket = s3.get_bucket(bucketName, validate=False)
    obj = bucket.get_key(key)
    if obj:
        return obj.exists()
    else:
        return False

def doesFileMatch(io1, io2):
    while True:
        b1 = io1.read(4096)
        b2 = io2.read(4096)
        if b1 == "":
            return b2 == ""
        elif b2 == "":
            return b1 == ""
        elif b1 != b2:
            return False

main()
